File: src/examples/shaping_tutorial/lesson_04/quest_sampling_shaper.cpp
#include "axom/config.hpp"
#include "axom/core.hpp"
#include "axom/slic.hpp"
#include "axom/primal.hpp"
#include "axom/sidre.hpp"
#include "axom/klee.hpp"
#include "axom/quest.hpp"
#include "axom/inlet.hpp"

#include "axom/fmt.hpp"
#include "axom/CLI11.hpp"

#include "mfem.hpp"

#ifdef AXOM_USE_MPI
  #include "mpi.h"
#endif

// NOTE: The shaping driver requires Axom to be configured with conduit and mfem
#if !defined(AXOM_USE_MFEM) && !defined(AXOM_USE_CONDUIT)
  #error Shaping functionality requires Axom to be configured with Conduit and MFEM
#endif

#include <iostream>
#include <limits>
#include <memory>
#include <string>
#include <vector>

namespace slic = axom::slic;
namespace klee = axom::klee;
namespace quest = axom::quest;
namespace primal = axom::primal;
namespace inlet = axom::inlet;
namespace sidre = axom::sidre;

// Mesh metadata (from examples, adapted)
struct MeshMetadata
{
  int dim;
  axom::Array<double> bb_min;
  axom::Array<double> bb_max;
  axom::Array<int> resolution;

  template <int DIM>
  axom::primal::BoundingBox<double, DIM> getBoundingBox() const
  {
    static_assert(DIM == 2 || DIM == 3, "Invalid dimension");
    SLIC_ASSERT_MSG(DIM == dim, "Template dimension must match MeshMetadata dimension");

    if constexpr(DIM == 2)
    {
      return axom::primal::BoundingBox<double, DIM>({bb_min[0], bb_min[1]}, {bb_max[0], bb_max[1]});
    }
    else
    {
      return axom::primal::BoundingBox<double, DIM>({bb_min[0], bb_min[1], bb_min[2]},
                                                    {bb_max[0], bb_max[1], bb_max[2]});
    }
  }

  static void defineSchema(inlet::Container& mesh_schema)
  {
    mesh_schema.addInt("dim", "Dimension (2 or 3)").required().range(2, 3);

    auto& bb = mesh_schema.addStruct("bounding_box", "Mesh bounding box").required();

    auto& min = bb.addStruct("min", "Minimum coordinates").required();
    min.addDouble("x", "Minimum x coordinate").required();
    min.addDouble("y", "Minimum y coordinate").required();
    min.addDouble("z", "Minimum z coordinate (only specify when dim is 3)");

    auto& max = bb.addStruct("max", "Maximum coordinates").required();
    max.addDouble("x", "Maximum x coordinate").required();
    max.addDouble("y", "Maximum y coordinate").required();
    max.addDouble("z", "Maximum z coordinate (only specify when dim is 3)");

    auto& res = mesh_schema.addStruct("resolution", "Mesh resolution").required();
    res.addInt("x", "Resolution in x direction").required().range(1, std::numeric_limits<int>::max());
    res.addInt("y", "Resolution in y direction").required().range(1, std::numeric_limits<int>::max());
    res.addInt("z", "Resolution in z direction (only specify when dim is 3)")
      .range(1, std::numeric_limits<int>::max());

    // Validate bounding box min/max
    bb.registerVerifier([](const inlet::Container& input) {
      bool valid = true;
      for(const std::string& axis : {"x", "y", "z"})
      {
        const std::string min_str = axom::fmt::format("min/{}", axis);
        const std::string max_str = axom::fmt::format("max/{}", axis);
        if(axis == "z" && (!input.contains(min_str) && !input.contains(max_str)))
        {
          continue;
        }

        if(const double min_val = input[min_str], max_val = input[max_str]; min_val >= max_val)
        {
          SLIC_WARNING(axom::fmt::format("Invalid bounding box range for {}-coordinate: {} >= {}",
                                         axis,
                                         min_val,
                                         max_val));
          valid = false;
        }
      }
      return valid;
    });

    // Validate presence/absence of z fields based on dim
    mesh_schema.registerVerifier([](const inlet::Container& input) {
      const int dim = input["dim"];
      bool valid = true;

      for(const auto& field : {"bounding_box/min/z", "bounding_box/max/z", "resolution/z"})
      {
        if(dim == 3)
        {
          if(!input.contains(field))
          {
            SLIC_WARNING(
              axom::fmt::format("Z-coordinate for '{}' is required when dimension is 3", field));
            valid = false;
          }
        }
        else if(dim == 2)
        {
          if(input.contains(field))
          {
            SLIC_WARNING(
              axom::fmt::format("Z-coordinate for '{}' should not be provided when dimension is 2",
                                field));
            valid = false;
          }
        }
      }

      return valid;
    });
  }
};

template <>
struct FromInlet<MeshMetadata>
{
  MeshMetadata operator()(const inlet::Container& input_data)
  {
    MeshMetadata result;

    result.dim = input_data["dim"];

    result.bb_min.resize(result.dim);
    result.bb_max.resize(result.dim);
    result.resolution.resize(result.dim);

    auto bb = input_data["bounding_box"];
    result.bb_min[0] = bb["min/x"];
    result.bb_min[1] = bb["min/y"];

    result.bb_max[0] = bb["max/x"];
    result.bb_max[1] = bb["max/y"];

    auto res = input_data["resolution"];
    result.resolution[0] = res["x"];
    result.resolution[1] = res["y"];

    if(result.dim == 3)
    {
      result.bb_min[2] = bb["min/z"];
      result.bb_max[2] = bb["max/z"];
      result.resolution[2] = res["z"];
    }

    return result;
  }
};

mfem::Mesh* createCartesianMesh(const MeshMetadata& meta, int outputOrder = 1)
{
  mfem::Mesh* mesh = nullptr;

  switch(meta.dim)
  {
  case 2:
  {
    const axom::NumericArray<int, 2> res {meta.resolution[0], meta.resolution[1]};
    const auto bbox = meta.getBoundingBox<2>();

    SLIC_INFO(axom::fmt::format("Creating 2D Cartesian mesh of res {} and bbox {}", res, bbox));
    mesh = quest::util::make_cartesian_mfem_mesh_2D(bbox, res, outputOrder);
  }
  break;
  case 3:
  {
    const axom::NumericArray<int, 3> res {meta.resolution[0], meta.resolution[1], meta.resolution[2]};
    const auto bbox = meta.getBoundingBox<3>();

    SLIC_INFO(axom::fmt::format("Creating 3D Cartesian mesh of res {} and bbox {}", res, bbox));
    mesh = quest::util::make_cartesian_mfem_mesh_3D(bbox, res, outputOrder);
  }
  break;
  default:
    SLIC_ERROR("Only 2D and 3D meshes are supported");
    break;
  }

#if defined(AXOM_USE_MPI) && defined(MFEM_USE_MPI)
  {
    int* partitioning = nullptr;
    int part_method = 0;
    mfem::Mesh* pmesh = new mfem::ParMesh(MPI_COMM_WORLD, *mesh, partitioning, part_method);
    delete[] partitioning;
    delete mesh;
    mesh = pmesh;
  }
#endif

  return mesh;
}

int main(int argc, char** argv)
{
  axom::utilities::raii::MPIWrapper mpi_raii_wrapper(argc, argv);
  axom::slic::SimpleLogger raii_logger;

  // --------------------------------------------------------------------------
  // CLI for input files
  // --------------------------------------------------------------------------
  axom::CLI::App app {"Shaping pipeline using separate Inlet mesh metadata and Klee shapes"};
  std::string inputFilename;  // Mesh metadata Inlet Lua
  std::string kleeFilename;   // Klee shape set YAML
  bool verbose = false;
  int outputOrder = 1;

  app.add_option("-m,--mesh_file", inputFilename)
    ->description("Mesh metadata Inlet Lua file")
    ->required()
    ->check(axom::CLI::ExistingFile);
  app.add_option("-k,--klee_file", kleeFilename)
    ->description("Klee shape set YAML file")
    ->required()
    ->check(axom::CLI::ExistingFile);
  app.add_flag("-v,--verbose", verbose)->description("Enable verbose (debug) logging");

  try
  {
    app.parse(argc, argv);
  }
  catch(const axom::CLI::ParseError& e)
  {
    int retval = app.exit(e);
#ifdef AXOM_USE_MPI
    MPI_Bcast(&retval, 1, MPI_INT, 0, MPI_COMM_WORLD);
#endif
    return retval;
  }

  slic::setAbortOnWarning(true);
  slic::setLoggingMsgLevel(verbose ? slic::message::Debug : slic::message::Info);

  // --------------------------------------------------------------------------
  // Parse and validate inlet input into MeshMetadata
  // --------------------------------------------------------------------------
  MeshMetadata meta = [&]() -> MeshMetadata {
    std::unique_ptr<inlet::Reader> reader = std::make_unique<inlet::LuaReader>();
    reader->parseFile(inputFilename);
    inlet::Inlet input(std::move(reader));

    auto& mesh_schema = input.addStruct("mesh", "Mesh metadata").required();
    MeshMetadata::defineSchema(mesh_schema);

    SLIC_ERROR_IF(!input.verify(), "Input validation failed.");

    // Parse mesh metadata and shaping params
    return input["mesh"].get<MeshMetadata>();
  }();

  // --------------------------------------------------------------------------
  // Set up computational mesh from MeshMetadata
  // --------------------------------------------------------------------------
  mfem::Mesh* mesh = createCartesianMesh(meta, outputOrder);
  constexpr bool dc_owns_data = true;  // Note: dc takes ownership of mesh
  sidre::MFEMSidreDataCollection dc("shaping", nullptr, dc_owns_data);
#ifdef AXOM_USE_MPI
  dc.SetMesh(MPI_COMM_WORLD, mesh);
#else
  dc.SetMesh(mesh);
#endif
  dc.SetMeshNodesName("positions");

  // --------------------------------------------------------------------------
  // Load and validate Klee shape set
  // --------------------------------------------------------------------------
  klee::ShapeSet shapeSet;
  try
  {
    shapeSet = klee::readShapeSet(kleeFilename);
  }
  catch(klee::KleeError& error)
  {
    std::vector<std::string> errs;
    for(const auto& verificationError : error.getErrors())
    {
      errs.push_back(axom::fmt::format(" - '{}': {}",
                                       static_cast<std::string>(verificationError.path),
                                       verificationError.message));
    }
    SLIC_WARNING(axom::fmt::format("Error parsing klee input:\n{}", axom::fmt::join(errs, "\n")));
    return 1;
  }

  // --------------------------------------------------------------------------
  // Setup shaper
  // --------------------------------------------------------------------------
  using RuntimePolicy = axom::runtime_policy::Policy;
  RuntimePolicy policy = RuntimePolicy::seq;

  auto shaper = std::make_unique<quest::SamplingShaper>(policy,
                                                        axom::policyToDefaultAllocatorID(policy),
                                                        shapeSet,
                                                        &dc);
  // TODO: Set some additional parameters....

  // --------------------------------------------------------------------------
  // Run pipeline
  // --------------------------------------------------------------------------
  // Process each shape
  SLIC_INFO(axom::fmt::format("{:=^80}", "Shaping"));
  for(const auto& shape : shapeSet.getShapes())
  {
    const std::string shapeFormat = shape.getGeometry().getFormat();
    SLIC_INFO_ROOT(
      axom::fmt::format("{:-^80}",
                        axom::fmt::format("Processing shape '{}' of material '{}' (format '{}')",
                                          shape.getName(),
                                          shape.getMaterial(),
                                          shapeFormat)));

    shaper->loadShape(shape);
    shaper->prepareShapeQuery(shape.getGeometry().getInputDimensions(), shape);
    shaper->runShapeQuery(shape);
    shaper->applyReplacementRules(shape);
    shaper->finalizeShapeQuery();
    slic::flushStreams();
  }

  return 0;
}

File: src/examples/shaping_tutorial/lesson_04/quest_sampling_shaper.py
#!/usr/bin/env python3

# Copyright (c) Lawrence Livermore National Security, LLC and other
# Axom Project Contributors. See top-level LICENSE and COPYRIGHT
# files for dates and other details.
#
# SPDX-License-Identifier: (BSD-3-Clause)
"""Shape a Klee input onto a computational mesh using Quest SamplingShaper."""

from argparse import ArgumentParser
from dataclasses import dataclass, field
from pathlib import Path

import pyklee
import pyinlet
import pyquest

MAX_INT = 2 ** 31 - 1


@dataclass
class MeshMetadata:
    dim: int = 2
    bb_min: list[float] = field(default_factory=lambda: [0.0, 0.0])
    bb_max: list[float] = field(default_factory=lambda: [1.0, 1.0])
    resolution: list[int] = field(default_factory=lambda: [10, 10])
    background_material: str = ""
    volume_fraction_order: int = 2
    mesh_order: int = 1
    sampling_resolution: int = 5
    sampling_method: pyquest.SamplingMethod = pyquest.SamplingMethod.InOut


def _make_reader(path: str):
    suffix = Path(path).suffix.lower()
    if suffix in {".yaml", ".yml"}:
        return pyinlet.YAMLReader()
    if suffix == ".lua" and hasattr(pyinlet, "LuaReader"):
        return pyinlet.LuaReader()

    supported = [".yaml", ".yml"]
    if hasattr(pyinlet, "LuaReader"):
        supported.append(".lua")

    raise RuntimeError(f"Unsupported mesh metadata extension for '{path}'. "
                       f"Expected {', '.join(supported[:-1])} or {supported[-1]}.")


def _define_mesh_schema(mesh_schema) -> None:
    mesh_schema.addInt("dim", "Dimension (2 or 3)").required().range(2, 3)

    bounding_box = mesh_schema.addStruct("bounding_box", "Mesh bounding box").required()

    minimum = bounding_box.addStruct("min", "Minimum coordinates").required()
    minimum.addDouble("x", "Minimum x coordinate").required()
    minimum.addDouble("y", "Minimum y coordinate").required()
    minimum.addDouble("z", "Minimum z coordinate (only specify when dim is 3)")

    maximum = bounding_box.addStruct("max", "Maximum coordinates").required()
    maximum.addDouble("x", "Maximum x coordinate").required()
    maximum.addDouble("y", "Maximum y coordinate").required()
    maximum.addDouble("z", "Maximum z coordinate (only specify when dim is 3)")

    resolution = mesh_schema.addStruct("resolution", "Mesh resolution").required()
    resolution.addInt("x", "Resolution in x direction").required().range(1, MAX_INT)
    resolution.addInt("y", "Resolution in y direction").required().range(1, MAX_INT)
    resolution.addInt("z",
                      "Resolution in z direction (only specify when dim is 3)").range(1, MAX_INT)

    mesh_schema.addString("background_material", "Optional background material")
    mesh_schema.addInt("volume_fraction_order",
                       "Order for volume fraction fields (>= 1)").range(1, MAX_INT)
    mesh_schema.addInt("mesh_order", "Order for mesh nodes (>= 1)").range(1, MAX_INT)
    mesh_schema.addInt("sampling_resolution", "Sampling resolution (>= 1)").range(1, MAX_INT)
    mesh_schema.addInt("quadrature_order",
                       "Legacy alias for sampling_resolution (>= 1)").range(1, MAX_INT)
    mesh_schema.addString("sampling_method", "Sampling method ('inout' or 'winding')").validValues(
        ["inout", "winding"])

    def verify_bounding_box(input_data) -> bool:
        valid = True
        for axis in ("x", "y", "z"):
            min_name = f"min/{axis}"
            max_name = f"max/{axis}"
            if axis == "z" and (not input_data.contains(min_name)
                                and not input_data.contains(max_name)):
                continue

            min_value = float(input_data[min_name])
            max_value = float(input_data[max_name])
            if min_value >= max_value:
                print(
                    f"Invalid bounding box range for {axis}-coordinate: {min_value} >= {max_value}")
                valid = False

        return valid

    def verify_dimension(input_data) -> bool:
        dim = int(input_data["dim"])
        valid = True

        for field_name in ("bounding_box/min/z", "bounding_box/max/z", "resolution/z"):
            if dim == 3:
                if not input_data.contains(field_name):
                    print(f"Z-coordinate for '{field_name}' is required when dimension is 3")
                    valid = False
            elif dim == 2 and input_data.contains(field_name):
                print(f"Z-coordinate for '{field_name}' should not be provided when dimension is 2")
                valid = False

        return valid

    bounding_box.registerVerifier(verify_bounding_box)
    mesh_schema.registerVerifier(verify_dimension)


def _mesh_metadata_from_inlet(mesh) -> MeshMetadata:
    metadata = MeshMetadata(dim=int(mesh["dim"]))
    bounding_box = mesh["bounding_box"]
    resolution = mesh["resolution"]

    metadata.bb_min = [float(bounding_box["min/x"]), float(bounding_box["min/y"])]
    metadata.bb_max = [float(bounding_box["max/x"]), float(bounding_box["max/y"])]
    metadata.resolution = [int(resolution["x"]), int(resolution["y"])]

    if metadata.dim == 3:
        metadata.bb_min.append(float(bounding_box["min/z"]))
        metadata.bb_max.append(float(bounding_box["max/z"]))
        metadata.resolution.append(int(resolution["z"]))

    if mesh.contains("background_material"):
        metadata.background_material = str(mesh["background_material"])
    if mesh.contains("volume_fraction_order"):
        metadata.volume_fraction_order = int(mesh["volume_fraction_order"])
    if mesh.contains("mesh_order"):
        metadata.mesh_order = int(mesh["mesh_order"])
    if mesh.contains("sampling_resolution"):
        metadata.sampling_resolution = int(mesh["sampling_resolution"])
    elif mesh.contains("quadrature_order"):
        metadata.sampling_resolution = int(mesh["quadrature_order"])

    if mesh.contains("sampling_method"):
        method = str(mesh["sampling_method"])
        metadata.sampling_method = (pyquest.SamplingMethod.WindingNumber
                                    if method == "winding" else pyquest.SamplingMethod.InOut)

    return metadata


def _read_mesh_metadata(path: str) -> MeshMetadata:
    reader = _make_reader(path)
    if not reader.parseFile(path):
        raise RuntimeError(f"Failed to parse '{path}'.")

    inlet = pyinlet.Inlet(reader)
    mesh_schema = inlet.addStruct("mesh", "Mesh metadata").required()
    _define_mesh_schema(mesh_schema)

    if not inlet.verify():
        raise RuntimeError("Mesh metadata validation failed.")

    return _mesh_metadata_from_inlet(inlet["mesh"])


def main() -> int:
    parser = ArgumentParser(
        description="Shaping pipeline using separate Inlet mesh metadata and Klee shapes.")
    parser.add_argument(
        "-m",
        "--mesh_file",
        required=True,
        help="Mesh metadata Inlet Lua or YAML file.",
    )
    parser.add_argument(
        "-k",
        "--klee_file",
        required=True,
        help="Klee shape-set YAML file.",
    )
    parser.add_argument("-v", "--verbose", action="store_true", help="Enable verbose logging.")
    args = parser.parse_args()

    try:
        meta = _read_mesh_metadata(args.mesh_file)
        shape_set = pyklee.readShapeSet(args.klee_file)
    except RuntimeError as exc:
        print(exc)
        return 1

    shaper = pyquest.SamplingShaper(
        shape_set,
        meta.dim,
        meta.bb_min,
        meta.bb_max,
        meta.resolution,
        meta.mesh_order,
    )
    shaper.setVerbosity(args.verbose)
    shaper.setSamplingResolution(meta.sampling_resolution)
    shaper.setVolumeFractionOrder(meta.volume_fraction_order)
    shaper.setSamplingMethod(meta.sampling_method)

    if meta.sampling_method == pyquest.SamplingMethod.InOut:
        shaper.setSamplesPerKnotSpan(50)

    if meta.background_material:
        shaper.importBackgroundMaterial(meta.background_material, meta.volume_fraction_order)

    for shape in shape_set.getShapes():
        shape_dim = shape.getGeometry().getInputDimensions()
        shaper.loadShape(shape)
        shaper.prepareShapeQuery(shape_dim, shape)
        shaper.runShapeQuery(shape)
        shaper.applyReplacementRules(shape)
        shaper.finalizeShapeQuery()

    shaper.adjustVolumeFractions()
    shaper.save()
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
